src/lang-promql/client/prometheus.ts

```diff
@@ -98,26 +98,59 @@ export interface PrometheusClient {
   labelValues(labelName: string, metricName?: string): Promise<string[]>;
 }
 
-interface APIResponse {
-  status: string;
-  data: any;
-  warnings: string[];
+interface APIResponse<T> {
+  status: 'success' | 'error';
+  data?: T;
+  error?: string;
+  warnings?: string[];
 }
 
+// These are status codes where the Prometheus API still returns a valid JSON body,
+// with an error encoded within the JSON.
+const badRequest = 400;
+const unprocessableEntity = 422;
+const serviceUnavailable = 503;
+
 // HTTPPrometheusClient is the HTTP client that should be used to get some information from the different endpoint provided by prometheus.
 export class HTTPPrometheusClient implements PrometheusClient {
   private readonly lookbackInterval = 60 * 60 * 1000 * 12; //12 hours
   private readonly url: string;
   private readonly cache: Cache;
   private readonly errorHandler?: (error: any) => void;
+  // For some reason, just assigning via "= fetch" here does not end up executing fetch correctly
+  // when calling it, thus the indirection via another function wrapper.
+  private readonly fetchFn: FetchFn = (input: RequestInfo, init?: RequestInit): Promise<Response> => fetch(input, init);
 
-  constructor(url: string, errorHandler?: (error: any) => void, lookbackInterval?: number) {
+  constructor(url: string, errorHandler?: (error: any) => void, lookbackInterval?: number, fetchFn?: FetchFn) {
     this.cache = new Cache();
     this.url = url;
     this.errorHandler = errorHandler;
     if (lookbackInterval) {
       this.lookbackInterval = lookbackInterval;
     }
+    if (fetchFn) {
+      this.fetchFn = fetchFn;
+    }
+  }
+
+  private fetchAPI<T>(resource: string): Promise<T> {
```

Comment:
but maybe with a different logic

Reply:
I abstracted away some of the error code / data handling in the `HTTPPrometheusClient` into its own function because the overall structure of each request is shared / more complex. This isn't really the case for the LSP, where the top-level JSON object returned is just the direct payload (like an array of completion items). But I should probably at least check error handling a bit more on the LSP side, on that...

Reply:
well at least handling the case when the server is returning an error 400. It happens sometimes

Reply:
Yup, done.